### PR TITLE
Geometry improvements

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR proposes to:

- change `slices` component of `CropBatch` to `locations`

- change format of each `location` to list of three `slices` (actual instances of `slice`)

- add `__getitem__` to SEG-Y and HDF5 geometries that allows for easier loads

- add `zero_traces` attribute for SEG-Y geometries prior to collecting spatial stats

### CODE REVIEW

File: scripts/horizon_assessment.py [hunk 1]
```diff
@@ -97,41 +97,44 @@
...
+                    'other_path': os.path.join(*other.path.split('/')[-2:]),
                     'average_l1': overlap_info['mean'],
+                    'other_coverage': other.coverage,
                 }
-                horizon.show(savepath=os.path.join(config['savedir'], f'{prefix}carcass_depthmap.png'))
+                other_prefix = 'carcass' if other.is_carcass else '@'
+                other.show(savepath=os.path.join(config['savedir'], f'{prefix}{other_prefix}_depthmap.png'))
 
-                cm = HorizonMetrics(carcass)
+                om = HorizonMetrics(other)
```
Comment (dmylzenova) [inline scripts/horizon_assessment.py:117]: That is an unclear two letters naming

File: seismiqb/src/crop_batch.py [hunk 2]
```diff
@@ -124,29 +127,33 @@ def crop(self, points, shape=None, loc=(0, 0, 0), side_view=False,
...
             as its upper rightmost point and must be cut from 'Cube.sgy' file.
-        shape : sequence
-            Desired shape of crops.
-        loc : sequence of numbers
-            Location of the point relative to the cut crop. Must be a location on unit cube.
+        shape : sequence, ndarray
+            Desired shape of crops along (iline, xline, height) axis. If ndarray, then must have the same length,
+            as `points`, and each row contains a shape for corresponding point.
+        direction : sequence of numbers
+            Direction of the cut crop relative to the point. Must be a vector on unit cube.
```
Comment (dmylzenova) [inline seismiqb/src/crop_batch.py:134]: Not very clear. Provide an example, please.

File: seismiqb/src/crop_batch.py [hunk 3]
```diff
@@ -502,68 +512,46 @@ def concat_components(self, ix, src, dst, axis=-1):
...
-            component of batch to store the resulting labels, o/w a storing object.
-        threshold : float
-            parameter of mask-thresholding.
-        averaging : str
-            method of pandas.groupby used for finding the center of a horizon.
-        coordinates : str
-            coordinates-mode to use for keys of point-cloud. Can be either 'cubic'
-            or 'lines'. In case of `lines`-option, `geometries` must be loaded as
-            a component of batch.
+            Component of batch to store the resulting labels, o/w a storing object.
```
Comment (dmylzenova) [inline seismiqb/src/crop_batch.py:527]: o/w? -otherwise? one way? output/write?

File: seismiqb/src/geometry.py [hunk 4]
```diff
@@ -100,21 +99,21 @@ class SeismicGeometry:
...
+    Refer to the documentation of respective classes to learn more about their structure, attributes and methods.
     """
     #TODO: add separate class for cube-like labels
     SEGY_ALIASES = ['sgy', 'segy', 'seg']
     HDF5_ALIASES = ['hdf5', 'h5py']
 
     # Attributes to store during SEG-Y -> HDF5 conversion
     PRESERVED = [
-        'depth', 'delay', 'sample_rate',
+        'depth', 'delay', 'sample_rate', 'cube_shape',
```
Comment (dmylzenova) [inline seismiqb/src/geometry.py:116]: I haven't found that, if you already haven't -  can you, please, also store line_cdp transform or a couple of values?

File: seismiqb/src/horizon.py [hunk 5]
```diff
@@ -1156,7 +1173,7 @@ def evaluate(self, compute_metric=True, supports=50, plot=True, savepath=None, p
 
 
     def check_proximity(self, other, offset=0):
-        """ Shortcut for :meth:`.HorizonMetrics.evaluate` to compare against the best match of list of horizons. """
+        """ Compute a number of stats on location of `self` relative to the `other` Horizons. """
```
Comment (dmylzenova) [inline seismiqb/src/horizon.py:1176]: If you wanted to make it more explanatory, you could do better.

File: seismiqb/src/geometry.py [hunk 6]
```diff
@@ -170,6 +170,40 @@ def __len__(self):
...
+            path_meta = os.path.splitext(self.path)[0] + '.hdf5'
+
+        with h5py.File(path_meta, "r") as file_meta:
+            for item in self.PRESERVED:
+                try:
+                    value = file_meta['/info/' + item][()]
+                    setattr(self, item, value)
+                    self.loaded.append(item)
+                except KeyError:
+                    pass
```
Comment (dmylzenova) [inline seismiqb/src/geometry.py:204]: Warnings, if some metadata is missing, wouldn't hurt.

File: seismiqb/src/geometry.py [hunk 7]
```diff
@@ -809,26 +903,15 @@ def process(self, **kwargs):
         No passing through data whatsoever.
         """
         _ = kwargs
-        # self.file_hdf5 = SafeIO(self.path, opener=h5pickle.File, mode='r')
-        self.file_hdf5 = h5pickle.File(self.path, mode='r')
+        self.file_hdf5 = SafeIO(self.path, opener=h5py.File, mode='r')
+        # self.file_hdf5 = h5py.File(self.path, mode='r')
```
Comment (dmylzenova) [inline seismiqb/src/geometry.py:907]: Why is it here?

**Review by dmylzenova — CHANGES_REQUESTED** [review]

File: seismiqb/src/crop_batch.py (hunk 2 above)
Comment (SergeyTsimfer) [reply seismiqb/src/crop_batch.py:134]: I was actually unable to do so. Do you still use this parameter in your researches?

File: seismiqb/src/horizon.py [hunk 8]
```diff
@@ -178,13 +178,15 @@ def add_to_mask(self, mask, locations=None, width=3, alpha=1, iterator=None, **k
         low = width // 2
         high = max(width - low, 0)
 
-        shift_1, shift_2, h_min = [np.min(item) for item in locations]
-        h_max = np.max(locations[-1])
+        shift_1, shift_2, h_min = [slc.start for slc in locations]
+        h_max = locations[-1].start
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply seismiqb/src/horizon.py:182]: Nice catch!

File: seismiqb/src/horizon.py [hunk 9]
```diff
@@ -951,7 +958,7 @@ def get_cube_values(self, window=23, offset=0, scale=False, chunk_size=256):
 
             # Subsequently add values from the cube to background, then shift horizon 1 unit lower
             for j in range(window):
-                background[idx_i, idx_x, np.full_like(heights, j)] = data_chunk[heights, idx_i, idx_x]
+                background[idx_i, idx_x, np.full_like(heights, j)] = data_chunk[idx_i, idx_x, heights]
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply seismiqb/src/horizon.py:961]: The caching of data along the horizon is on the `Metrics` side, as it always was.

File: seismiqb/src/crop_batch.py (hunk 2 above)
Comment (dmylzenova) [reply seismiqb/src/crop_batch.py:134]: Now I don't

File: seismiqb/src/geometry.py [hunk 10]
```diff
@@ -637,6 +644,25 @@ def collect_stats(self, spatial=True, bins=25, num_keep=5000, **kwargs):
         self.has_stats = True
         self.store_meta()
 
+    def add_rotation_matrix(self):
+        """ Add transform from INLINE/CROSSLINE corrdinates to CDP system. """
```
Comment (dmylzenova) [inline seismiqb/src/geometry.py:647]: A usage example won't hurt

**Pull request merged**